src/aes_key.c

```diff
@@ -1029,6 +1073,105 @@ __aes_key_reset(struct zpc_aes_key *aes_key)
 	aes_key->refcount = 1;
 }
 
+u16 aesprotkeylen_from_pvsectype(u16 pvsectype)
+{
+	switch (pvsectype) {
+	case ZPC_AES_SECRET_AES_128:
+		return 16 + 32;
+	case ZPC_AES_SECRET_AES_192:
+		return 24 + 32;
+	case ZPC_AES_SECRET_AES_256:
+		return 32 + 32;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+/*
+ * Verify that a given pvsecret ID is a valid ID on this system, i.e. an UV
+ * secret exists with this ID and has the expected key length.
+ */
+int aes_key_blob_is_valid_pvsecret_id(struct zpc_aes_key *aes_key, const unsigned char *id)
+{
+	struct pkey_verifykey2 io;
+	unsigned char buf[sizeof(struct uvrsecrettoken)];
+	struct uvrsecrettoken *clrtok = (struct uvrsecrettoken *)&buf;
+	int rc;
+
+	memset(buf, 0, sizeof(buf));
+	clrtok->version = TOKVER_UV_SECRET;
+	switch (aes_key->keysize) {
+	case 128:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_128;
+		break;
+	case 192:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_192;
+		break;
+	default:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_256;
+		break;
+	}
+	clrtok->secret_len = aesprotkeylen_from_pvsectype(clrtok->secret_type);
+	memcpy(clrtok->secret_id, id, UV_SECRET_ID_LEN);
+
+	memset(&io, 0, sizeof(io));
+	io.key = buf;
+	io.keylen = sizeof(struct uvrsecrettoken);
+
+	rc = ioctl(pkeyfd, PKEY_VERIFYKEY2, &io);
+	if (rc != 0) {
+		return ZPC_ERROR_IOCTLVERIFYKEY2;
+	}
+
+	return 0;
+}
+
+/*
+ * (Re)derive protected key from a retrievable secret ID.
+ * Caller must hold aes_key's wr lock.
+ */
+int aes_key_pvsec2prot(struct zpc_aes_key *aes_key)
+{
+	struct pkey_kblob2pkey3 io;
+	unsigned char buf[sizeof(struct uvrsecrettoken)];
+	struct uvrsecrettoken *clrtok = (struct uvrsecrettoken *)&buf;
+	int rc;
+
+	memset(buf, 0, sizeof(buf));
+	clrtok->version = TOKVER_UV_SECRET;
+	switch (aes_key->keysize) {
+	case 128:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_128;
+		break;
+	case 192:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_192;
+		break;
+	default:
+		clrtok->secret_type = ZPC_AES_SECRET_AES_256;
+		break;
+	}
+	clrtok->secret_len = aesprotkeylen_from_pvsectype(clrtok->secret_type);
+	memcpy(clrtok->secret_id, aes_key->cur.sec, UV_SECRET_ID_LEN);
+
+	memset(&io, 0, sizeof(io));
+	io.key = buf;
+	io.keylen = sizeof(struct uvrsecrettoken);
+	io.apqns = NULL;
```

Comment:
setting apqns and apqn_entries to zero /NULL is superfluous, you are memsetting the struct above anyway.

BTW: This function is very similar as the above function. can this be refactored into a common routine maybe?

Reply:
Removed the NULL/0 settings. Still thinking about some refactoring ...

Reply:
Now did the refactoring, see new subroutine aes_key_make_uvrsecrettoken()